http-client/src/main/java/io/airlift/http/client/ScatteringByteArrayInputStream.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2010 Proofpoint, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.airlift.http.client;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Iterator;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.Math.min;
+import static java.lang.System.arraycopy;
+import static java.util.Objects.requireNonNull;
+
+public class ScatteringByteArrayInputStream
```

Comment:
This class should be named `Gathering` since it is gathering the data from multiple buffers into a single input stream. Scattering is used for reads when a single read is scattered into multiple output buffers.